src/main/java/org/jetbrains/bsp/bazel/server/BepServer.java

```diff
@@ -88,156 +87,122 @@ public void publishLifecycleEvent(
   @Override
   public StreamObserver<PublishBuildToolEventStreamRequest> publishBuildToolEventStream(
       StreamObserver<PublishBuildToolEventStreamResponse> responseObserver) {
-    return new StreamObserver<PublishBuildToolEventStreamRequest>() {
-
-      @Override
-      public void onNext(PublishBuildToolEventStreamRequest request) {
-        if (request
-            .getOrderedBuildEvent()
-            .getEvent()
-            .getBazelEvent()
-            .getTypeUrl()
-            .equals("type.googleapis.com/build_event_stream.BuildEvent")) {
-          handleEvent(request.getOrderedBuildEvent().getEvent());
-        } else {
-          //          System.out.println("Got this request " + request);
-        }
-        PublishBuildToolEventStreamResponse response =
-            PublishBuildToolEventStreamResponse.newBuilder()
-                .setStreamId(request.getOrderedBuildEvent().getStreamId())
-                .setSequenceNumber(request.getOrderedBuildEvent().getSequenceNumber())
-                .build();
-        responseObserver.onNext(response);
+    return new BepStreamObserver(this, responseObserver);
+  }
+
+  public TreeSet<Uri> getCompilerClasspath() {
+    return compilerClasspath;
+  }
+
+  public Map<String, String> getDiagnosticsProtosLocations() {
+    return diagnosticsProtosLocations;
+  }
+
+  public void collectDiagnostics(
+      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics,
+      BuildTargetIdentifier target,
+      String diagnosticsLocation)
+      throws IOException {
+    diagnosticsDispatcher.collectDiagnostics(filesToDiagnostics, target, diagnosticsLocation);
+  }
+
+  public void emitDiagnostics(
+      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics, BuildTargetIdentifier target) {
+    diagnosticsDispatcher.emitDiagnostics(filesToDiagnostics, target);
+  }
+
+  public void handleEvent(BuildEvent buildEvent) {
+    try {
+      BuildEventStreamProtos.BuildEvent event =
+          BuildEventStreamProtos.BuildEvent.parseFrom(buildEvent.getBazelEvent().getValue());
+
+      LOGGER.info("Got event" + event + "\nevent-end\n");
+
+      if (event.hasStarted() && event.getStarted().getCommand().equals("build")) {
```

Comment:
I think that this methods could be changed a bit:
```
if (event.hasStarted() && event.getStarted().getCommand().equals("build")) {
  processedStartedEvent(event.getStarted());
}
```

could be moved to the separate  method with correct name

Reply:
Thought about it but couldn't find a good name and forgot it. Do you have something better than hasBuildStarted(event)?

Reply:
good enough for me, but I was thinking about an entire if section, as well for similar below

Reply:
it applies to if sections below too

Reply:
You mean wrapping the if in something like _tryProcessFinishedEvent_?

Reply:
try can suggest `try - catch` block but in general yes:
```
foo(event) {
  if (event.hasStarted() && event.getStarted().getCommand().equals("build")) {
    processedStartedEvent(event.getStarted());
  }
}
``` 
and then replace this code with foo(event)